tests/realtikvtest/sessiontest/session_fail_test.go

```diff
@@ -317,3 +320,41 @@ func TestTiKVClientReadTimeout(t *testing.T) {
 	explain = fmt.Sprintf("%v", rows[0])
 	require.Regexp(t, ".*TableReader.* root  time:.*, loops:.* cop_task: {num: 1, .* rpc_num: (3|4|5).*", explain)
 }
+
+func TestBatchClientDataRace(t *testing.T) {
+	// This test uses to detect data race, so it should run by `go test -race`.
+	if !*realtikvtest.WithRealTiKV {
+		t.Skip("skip test since it's only work for tikv")
+	}
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.TiKVClient.MaxBatchSize = 128
+	})
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("create table if not exists t (a int primary key, b int)")
+
+	var wg sync.WaitGroup
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			tk := testkit.NewTestKit(t, store)
+			tk.MustExec("use test")
+			for j := 0; j < 100; j++ {
+				ctx, cancel := context.WithCancel(context.Background())
+				go func() {
+					// mock for kill query execution or timeout.
+					time.Sleep(time.Millisecond * time.Duration(rand.Intn(10)+1))
+					cancel()
+				}()
+				rs, _ := tk.ExecWithContext(ctx, "select * from t where a = 1")
+				if rs != nil {
+					session.ResultSetToStringSlice(ctx, tk.Session(), rs)
+				}
+			}
+
```

Comment:
Please remove this useless line.
Besides, I ran this test（with `--with-real-tikv` and `-race`） locally with the master version 2 twice and passed both times. Is this test unreliable or does it need to work well on CI?

Reply:
Nice catch. I wrote this test in the release-6.5 branch first time, and the test will fail in release-6.5 branch.

Master branch can pass the test even without https://github.com/tikv/client-go/pull/1021 , I still keep the test to detect data race in future.

Reply:
In release-6.5 run this test will fail.
```sh
github.com/pingcap/tidb  release-6.5 ✗                                                                                                                                          1d23h ⚑
▶ cd tests/realtikvtest/sessiontest/
(base)
tests/realtikvtest/sessiontest  release-6.5 ✗                                                                                                                                   1d23h ⚑
▶ go test -race --tags=intest -run=TestBatchClientDataRace --with-real-tikv=true
...
...
...
[2023/10/18 14:56:15.607 +08:00] [INFO] [pd.go:209] ["closed pd http client"]
--- FAIL: TestBatchClientDataRace (31.54s)
    testing.go:1465: race detected during execution of test
FAIL
exit status 1
FAIL    github.com/pingcap/tidb/tests/realtikvtest/sessiontest  32.724s
```